parquet/src/main/scala/magnolify/parquet/ParquetField.scala

```diff
@@ -0,0 +1,438 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package magnolify.parquet
+
+import magnolia1._
+import magnolify.shared.{Converter => _, _}
+import magnolify.shims.FactoryCompat
+
+import java.nio.{ByteBuffer, ByteOrder}
+import java.time.LocalDate
+import java.util.UUID
+import org.apache.parquet.io.ParquetDecodingException
+import org.apache.parquet.io.api._
+import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName
+import org.apache.parquet.schema.Type.Repetition
+import org.apache.parquet.schema.{LogicalTypeAnnotation, Type, Types}
+
+import scala.annotation.implicitNotFound
+import scala.collection.concurrent
+import scala.language.experimental.macros
+import scala.collection.compat._
+
+sealed trait ParquetField[T] extends Serializable {
+
+  @transient private lazy val schemaCache: concurrent.Map[UUID, Type] =
+    concurrent.TrieMap.empty
+
+  protected def buildSchema(cm: CaseMapper): Type
+  def schema(cm: CaseMapper): Type =
+    schemaCache.getOrElseUpdate(cm.uuid, buildSchema(cm))
+
+  val hasAvroArray: Boolean = false
+  val fieldAnnotations: Map[String, String]
+  val typeAnnotation: Option[String]
+  protected val isGroup: Boolean = false
+  protected def isEmpty(v: T): Boolean
+  def write(c: RecordConsumer, v: T)(cm: CaseMapper): Unit
+  def newConverter: TypeConverter[T]
+
+  protected def writeGroup(c: RecordConsumer, v: T)(cm: CaseMapper): Unit = {
+    if (isGroup) {
+      c.startGroup()
+    }
+    write(c, v)(cm)
+    if (isGroup) {
+      c.endGroup()
+    }
+  }
+}
+
+object ParquetField {
+  type Typeclass[T] = ParquetField[T]
+
+  sealed trait Record[T] extends ParquetField[T] {
+    override protected val isGroup: Boolean = true
+    override protected def isEmpty(v: T): Boolean = false
+  }
+
+  def join[T](caseClass: CaseClass[Typeclass, T]): Record[T] = new Record[T] {
+    override def buildSchema(cm: CaseMapper): Type =
+      caseClass.parameters
+        .foldLeft(Types.requiredGroup()) { (g, p) =>
+          g.addField(Schema.rename(p.typeclass.schema(cm), cm.map(p.label)))
+        }
+        .named(caseClass.typeName.full)
+
+    override val hasAvroArray: Boolean = caseClass.parameters.exists(_.typeclass.hasAvroArray)
+
+    override val fieldAnnotations: Map[String, String] =
+      caseClass.parameters.flatMap { param =>
+        val label = param.label
+        val nestedAnnotations = param.typeclass.fieldAnnotations.map { case (k, v) =>
+          (s"$label.$k", v)
```

Comment:
Prefer arrow notations for maps
```suggestion
          s"$label.$k" -> v
```

Reply:
Done